File: lib/stream.dart
// crypto-fl: cryptography primitives and wrappers
// Copyright 2026 Dark Bio AG. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

/// Streaming authenticated encryption based on age's STREAM construction.
///
/// https://eprint.iacr.org/2015/189.pdf
///
/// Plaintext is split into 64 KiB chunks, each sealed with ChaCha20-Poly1305
/// under a nonce that counts up and marks the final chunk, so truncation and
/// reordering are detected. Both functions take and return whole buffers.
///
/// ```dart
/// import 'dart:convert';
///
/// import 'package:darkbio_crypto/rand.dart' as rand;
/// import 'package:darkbio_crypto/stream.dart' as stream;
///
/// void example() {
///   // Never reuse a key across streams
///   final key = rand.bytes(32);
///
///   final ciphertext = stream.encrypt(key: key, plaintext: utf8.encode('hello stream'));
///   final plaintext = stream.decrypt(key: key, ciphertext: ciphertext);
///   assert(utf8.decode(plaintext) == 'hello stream');
/// }
/// ```
library;

import 'dart:typed_data';

import 'src/generated/api/stream.dart' as ffi;

/// Encrypts [plaintext] as a STREAM under the 32-byte [key].
///
/// The key must **never** be repeated across multiple streams. Derive it with
/// HKDF from both a random file key and a random nonce.
///
/// Throws if [key] is not 32 bytes long.
Uint8List encrypt({required Uint8List key, required Uint8List plaintext}) =>
    ffi.streamEncrypt(key: key, plaintext: plaintext);

/// Decrypts a STREAM [ciphertext] under the 32-byte [key] it was encrypted
/// with.
///
/// Throws if [key] is not 32 bytes long, or if the ciphertext does not
/// authenticate under it, such as when it was truncated, reordered or
/// tampered with.
Uint8List decrypt({required Uint8List key, required Uint8List ciphertext}) =>
    ffi.streamDecrypt(key: key, ciphertext: ciphertext);

File: rust/src/api/stream.rs
use flutter_rust_bridge::frb;
use std::io::{Read, Write};

/// Encrypts plaintext using the STREAM construction with ChaCha20-Poly1305.
/// The key must be exactly 32 bytes and should never be reused across streams.
#[frb(sync)]
pub fn stream_encrypt(key: Vec<u8>, plaintext: Vec<u8>) -> Result<Vec<u8>, String> {
    let key_array: [u8; 32] = key
        .try_into()
        .map_err(|_| "Invalid key length, expected 32 bytes".to_string())?;

    let payload_key = darkbio_crypto::stream::PayloadKey(key_array.into());

    let mut ciphertext = Vec::new();
    let mut writer = darkbio_crypto::stream::Stream::encrypt(payload_key, &mut ciphertext);

    writer
        .write_all(&plaintext)
        .map_err(|e| format!("Encryption write error: {}", e))?;

    writer
        .finish()
        .map_err(|e| format!("Encryption finish error: {}", e))?;

    Ok(ciphertext)
}

/// Decrypts ciphertext using the STREAM construction with ChaCha20-Poly1305.
/// The key must be exactly 32 bytes and must match the key used for encryption.
#[frb(sync)]
pub fn stream_decrypt(key: Vec<u8>, ciphertext: Vec<u8>) -> Result<Vec<u8>, String> {
    let key_array: [u8; 32] = key
        .try_into()
        .map_err(|_| "Invalid key length, expected 32 bytes".to_string())?;

    let payload_key = darkbio_crypto::stream::PayloadKey(key_array.into());

    let mut reader = darkbio_crypto::stream::Stream::decrypt(payload_key, ciphertext.as_slice());

    let mut plaintext = Vec::new();
    reader
        .read_to_end(&mut plaintext)
        .map_err(|e| format!("Decryption error: {}", e))?;

    Ok(plaintext)
}
